server.py

```diff
@@ -1,36 +1,376 @@
 
-app.static('/', 'build/')
+app.static("/", "build/")
+
+
+# JSON type for coach data.
+CoachJSON = Dict[str, Union[str, int, bool, float, "CoachJSON"]]
+
+# ------------------------------------------------------------------------------
+# Database
+# ------------------------------------------------------------------------------
+
+
+class Coach(
+    namedtuple(
+        "_Coach",
+        [
+            "id",
+            "name",
+            "bio",
+            "available",
+            "birth_year",
+            "gender",
+            "languages",
+            "need",
+            "rights",
+            "housing",
+        ],
+    )
+):
+    """A coach object - an entry in the database."""
+
+    def __new__(
+        cls,
+        *,
+        id: Optional[int] = None,
+        name: str,
+        bio: str,
+        available: bool = True,
+        birth_year: int,
+        gender: str,
+        languages: Dict[str, int],
+        need: Collection[int],
+        rights: Collection[int],
+        housing: Collection[int],
+    ):
+        # TODO: Validate args.
+        return super().__new__(
+            cls,
+            id=id,
+            name=name,
+            bio=bio,
+            available=available,
+            birth_year=birth_year,
+            gender=gender,
+            languages=languages,
+            need=need,
+            rights=rights,
+            housing=housing,
+        )
+
+    def to_json(self) -> CoachJSON:
+        return self._asdict()
 
-def check_request_for_authorization_status(request):
-    sanic.log.logger.info(f"Checking authorization: IP is {request.ip}")
-    match = (request.ip == "127.0.0.1")
-    sanic.log.logger.info(f"Check IP matches 127.0.0.1: {match}")
+    @classmethod
+    def from_json(cls, obj: CoachJSON) -> "Coach":
+        return cls(**obj)
 
-    return match
+    def copy(self) -> "Coach":
+        return self.from_json(self.to_json())
 
-def authorized(f):
-    @functools.wraps(f)
-    async def decorated_function(request, *args, **kwargs):
-        is_authorized = check_request_for_authorization_status(request)
 
-        if is_authorized:
-            # Run the handler method and return the response.
-            response = await f(request, *args, **kwargs)
-            return response
+class CoachDB:
+    """The in-memory coach database."""
+
+    def __init__(
+        self, file: Optional[os.PathLike] = None, *, auto_persist: bool = True
+    ):
+        """
+        :param file:
+            File to use as persistent storage of the DB. If not found it will be
+            created.
+        :param auto_persist:
+            Whether to automatically persist the DB to file on operations that
+            modify the DB. Ignored if file is not given.
+            Set this to False if manual batching is desired.
+        """
+        self._db: Dict[int, Coach] = dict()
+        self._file = file
+        self._auto_persist = auto_persist and file
+        self._next_id: int = 0
```

Comment:
When we initialise the in-memory database, could we read in the file first and then set self._next_id to be the next ID on from those in the database? This would save asserting later on, and allow restarting the server.

Reply:
I'm not sure I follow. Perhaps you missed the fact that 'next_id' is stored in the JSON file? See [CoachDB.persist()](https://github.com/richcooper95/crisis/pull/8/files/564275d5eaca1a91daee7a452a40c8088d553e5e#diff-d355af669766ee3562643e8e5034f715R158).

When loading from file there was a few options w.r.t. allocating coach IDs:
 - Start back at 0 and skip any IDs that were already in the DB when creating new entries
 - Start 1 higher than the max ID in the JSON-persisted DB
 - Store the 'next ID' in the JSON file (as metadata to the DB data) and restore it when reading in the file

I went for the 3rd option because this is the only one that avoids a new coach being assigned an ID of a previously removed coach (think of the case where the latest added coach was removed for option 2).

W.r.t. setting `self._next_id = 0` here, I think this is quite a robust way to do initialisation: "initialise everything and then overwrite anything that needs special handling", such that things have more of a chance of continuing to work if the special handling changes :)

Reply:
Yep, I did miss that - apologies! All sounds good 👍